patches/server/0504-Add-BlockPreDispenseEvent.patch

```diff
@@ -16,6 +16,18 @@ index 5593a0aa9e618071b6521b213dde0f628348c3dc..644e64850479cea20a98b8a65503ccf3
                      DispenserBlock.eventFired = false; // CraftBukkit - reset event status
                      tileentitydispenser.setItem(i, idispensebehavior.dispense(sourceblock, itemstack));
                  }
+diff --git a/src/main/java/net/minecraft/world/level/block/DropperBlock.java b/src/main/java/net/minecraft/world/level/block/DropperBlock.java
+index 1d13f8a1009d6eda351c697052d499d594a6aaa8..7cc7ee9d9cb0dd1b785e1b48a1296fd914ba14a4 100644
+--- a/src/main/java/net/minecraft/world/level/block/DropperBlock.java
++++ b/src/main/java/net/minecraft/world/level/block/DropperBlock.java
+@@ -65,6 +65,7 @@ public class DropperBlock extends DispenserBlock {
+                 ItemStack itemstack = tileentitydispenser.getItem(i);
+ 
+                 if (!itemstack.isEmpty()) {
++                    if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockPreDispenseEvent(world, pos, itemstack, i)) return; // Paper - Add BlockPreDispenseEvent
```

Comment:
This would fire the event when the dropper is moving the item into a different inventory. Do we want that? Or instead should this be fired directly before it's actually dispensed (which happens a few lines below this one).

I'm not 100% sure, but either way, it should be clarified.

Reply:
Yeah, I think it's sensible to call the pre dispense event right before the dispense behaviour is invoked. For the rest of the logic between inventories there is the inventory move item event